File: specify/static/js/uiformatters.js
define([
    'jquery', 'underscore', 'backbone',
    'text!context/app.resource?name=UIFormatters!noinline'
], function($, _, Backbone, xml) {
    "use strict";
    var uiformatters =  $($.parseXML(xml));

    function UIFormatter(fields) {
        this.fields = fields;
    }
    UIFormatter.extend = Backbone.Model.extend;
    _(UIFormatter.prototype).extend({
        value: function() {
            return _(this.fields).pluck('value').join('');
        },
        parseRegexp: function() {
            return '^' + _(this.fields).map(function(field) {
                return '(' + field.wildOrValueRegexp() + ')';
            }).join('') + '$';
        },
        parse: function(value) {
            var match = RegExp(this.parseRegexp()).exec(value);
            match && match.shift();
            return match;
        },
        canAutonumber: function() {
            return _.any(_.invoke(this.fields, 'canAutonumber'));
        },
        needsAutoNumber: function(values) {
            return _.any(this.fields, function(field, i) {
                return field.isWild(values[i]);
            });
        },
        canonicalize: function(values) {
            return _.map(this.fields, function(field, i) {
                    return field.canonicalize(values[i]);
            }).join('');
        }
    });

    function Field(options) {
        this.size = options.size;
        this.value = options.value;
        this.inc = options.inc;
        this.byYear = options.byYear;
    }
    Field.extend = Backbone.Model.extend;
    _.extend(Field.prototype, {
        canAutonumber: function() {
            return this.inc || this.byYear;
        },
        wildRegexp: function() {
            return escapeRegExp(this.value);
        },
        isWild: function(value) {
            return RegExp(this.wildRegexp()).test(value) &&
                !RegExp(this.valueRegexp()).test(value);
        },
        wildOrValueRegexp: function() {
            return this.canAutonumber() ? this.wildRegexp() + '|' + this.valueRegexp()
                : this.valueRegexp();
        },
        canonicalize: _.identity
    });
    Field.forNode = function(node) {
        node = $(node);
        return new ({
            'year': YearField,
            'numeric': NumericField,
            'alphanumeric': AlphaNumField,
            'separator': SeparatorField
        }[node.attr('type')])({
            size: parseInt(node.attr('size'), 10),
            value: node.attr('value'),
            inc: node.attr('inc') === 'true',
            byYear: node.attr('byyear') === 'true'
        });
    };

    var NumericField = Field.extend({
        constructor: function(options) {
            options.value = Array(options.size+1).join('#');
            Field.call(this, options);
        },
        valueRegexp: function() {
            return '\\d{' + this.size + '}';
        }
    });

    var YearField = Field.extend({
        valueRegexp: function() {
            return '\\d{' + this.size + '}';
        }
    });

    var AlphaNumField = Field.extend({
        valueRegexp: function() {
            return '[a-zA-Z0-9]{' + this.size + '}';
        }
    });

    var SeparatorField = Field.extend({
        isWild: function() { return false; },
        valueRegexp: Field.prototype.wildRegexp
    });

    var CatalogNumberNumeric = UIFormatter.extend({
        constructor: function() {
            UIFormatter.call(this,  [
                new CatalogNumberNumeric.Field({ size: 9, inc: true })
            ]);
        }
    }, {
        Field: NumericField.extend({
            valueRegexp: function() {
                return '\\d{0,' + this.size + '}';
            },
            canonicalize: function(value) {
                return Array(this.size - value.length + 1).join('0') + value;
            }
        }
    )});


    function escapeRegExp(str) {
        return str.replace(/[-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, "\\$&");
    }

    function getUIFormatter(name) {
        var node = $(uiformatters.find('[name="' + name + '"]'));
        if (!node) return null;
        var external = node.find('external');
        if (external.length) {
            switch ($(external).text().split('.').pop()) {
            case 'CatalogNumberUIFieldFormatter':
                return new CatalogNumberNumeric();
            default:
                return null;
            }
        } else {
            return new UIFormatter(
                _(node.find('field')).map(Field.forNode));
        }
    }

    return {
        getByName: getUIFormatter,
        UIFormatter: UIFormatter,
        Field: Field
    };
});

File: specify/uiformatters.py
import re
from xml.etree import ElementTree
from datetime import date

from filter_by_col import filter_by_collection
from context.app_resource import get_app_resource

def get_uiformatter(collection, user, formatter_name):
    xml, __ = get_app_resource(collection, user, "UIFormatters")
    node = ElementTree.XML(xml).find('.//format[@name="%s"]' % formatter_name)
    return node and UIFormatter(node)

class AutonumberOverflowException(Exception):
    pass

class UIFormatter(object):
    def __init__(self, node):
        self.system = node.attrib['system'] == 'true'
        self.name = node.attrib['name']
        self.model_name = node.attrib['class'].split('.')[-1]
        self.field_name = node.attrib['fieldname']
        self.is_external = node.find('external') is not None
        if self.is_external: return
        self.fields = map(new_field, node.findall('field'))

    def parse_regexp(self):
        regexp = ''.join('(%s)' % f.wild_or_value_regexp() for f in self.fields)
        return '^%s$' % regexp

    def parse(self, value):
        match = re.match(self.parse_regexp(), value)
        if match is None:
            raise ValueError("value doesn't match formatter")
        return match.groups()

    def autonumber_regexp(self, vals):
        regexp = []
        for field, val in zip(self.fields, vals):
            if field.is_wild(val):
                regexp.append('(%s)' % field.value_regexp())
            else:
                regexp.append('(%s)' % re.escape(val))
        return '^%s$' % ''.join(regexp)

    def fillin_year(self, vals, year=None):
        filled_vals = []
        for field, val in zip(self.fields, vals):
            if field.by_year and field.is_wild(val):
                filled_vals.append("%d" % (year or date.today().year))
            else:
                filled_vals.append(val)
        return filled_vals

    def autonumber(self, collection, model, fieldname, vals, year=None):
        with_year = self.fillin_year(vals, year)

        objs = model.objects.filter(**{ fieldname + '__regex': self.autonumber_regexp(with_year) })
        try:
            biggest = filter_by_collection(objs, collection).order_by('-' + fieldname)[0]
        except IndexError:
            filled_vals = self.fill_vals_no_prior(with_year)
        else:
            filled_vals = self.fill_vals_after(getattr(biggest, fieldname))
        return ''.join(filled_vals)

    def fill_vals_after(self, prior):
        filled = []
        for field, val in zip(self.fields, self.parse(prior)):
            if field.inc:
                new_val = ("%0" + str(field.size) + "d") % (1 + int(val))
                if len(new_val) > field.size:
                    raise AutonumberOverflowException()
                filled.append(new_val)
            else:
                filled.append(val)
        return filled

    def fill_vals_no_prior(self, vals):
        filled = []
        for field, val in zip(self.fields, vals):
            if field.is_wild(val):
                filled.append(self.size * "0")
            else:
                filled.append(val)
        return filled

def new_field(node):
    by_type = {
        'numeric': NumericField,
        'year': YearField,
        'alphanumeric': AlphaNumField,
        'separator': SeparatorField
        }
    return by_type[node.attrib['type']](node)

class Field(object):
    def __init__(self, node):
        self.type = node.attrib['type']
        self.size = int(node.attrib['size'])
        self.value = node.attrib.get('value')
        self.inc = node.attrib.get('inc') == 'true'
        self.by_year = node.attrib.get('byyear') == 'true'

    def can_autonumber(self):
        return self.inc or self.by_year

    def wild_regexp(self):
        return re.escape(self.value)

    def is_wild(self, value):
        return re.match(self.wild_regexp(), value) \
            and not re.match(self.value_regexp(), value)

    def wild_or_value_regexp(self):
        if self.can_autonumber():
            return '%s|%s' % (self.wild_regexp(), self.value_regexp())
        else:
            return self.value_regexp()

class NumericField(Field):
    def __init__(self, node):
        Field.__init__(self, node)
        self.value = self.size * '#'

    def value_regexp(self):
        return r'[0-9]{%d}' % self.size

class YearField(Field):
    def value_regexp(self):
        return r'[0-9]{%d}' % self.size

class AlphaNumField(Field):
    def value_regexp(self):
        return r'[a-zA-Z0-9]{%d}' % self.size

class SeparatorField(Field):
    def is_wild(self, value):
        return False

    def value_regexp(self):
        return self.wild_regexp()
